File: solutions/java/binary-search-tree/1/src/main/java/BinarySearchTree.java
import java.util.ArrayList;
import java.util.List;

class BinarySearchTree<T extends Comparable<T>> {
    private Node<T> root;
    private List<T> asSortedList;
    private List<T> asLevelSortedList;    

    void insert(T value) {        
        if (this.root == null){
            this.root = new Node<>(value);
            asLevelSortedList = new ArrayList<>();
        }else {
            this.root.insert(value);
        }
        asLevelSortedList.add(value);
    }

    List<T> getAsSortedList() {
        asSortedList = new ArrayList<>();
        shutSortedList();
        return asSortedList;
    }

    void shutSortedList(){
         sortedList(root);
    }
    void sortedList(Node<T> data){
        if (data == null) return;
        else {
            sortedList(data.getLeft());
            asSortedList.add(data.getData());
            sortedList(data.getRight());
        }
    }
    List<T> getAsLevelOrderList() {
        return asLevelSortedList;
    }

    Node<T> getRoot() {
        return root;
    }

    static class Node<T extends Comparable<T>> {

        T data;
        Node<T> left;
        Node<T> right;

        public Node(T data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }

        Node<T> getLeft() {
            return left;
        }

        Node<T> getRight() {
            return right;
        }

        T getData() {
            return data;
        }

        void insert(T data){
            if(data.compareTo(this.data) <= 0){
                if (this.left == null){
                    this.left = new Node<>(data);
                }else {
                    this.left.insert(data);
                }
            }else {
                if (this.right == null){
                    this.right = new Node<>(data);
                }else {
                    this.right.insert(data);
                }
            }
        }
    }
}

File: solutions/python/binary-search-tree/1/binary_search_tree.py
class TreeNode:
    def __init__(self, data, left=None, right=None):
        self.data = data
        self.left = left
        self.right = right

    def __str__(self):
        return f'TreeNode(data={self.data}, left={self.left}, right={self.right})'

    def insert(self, data):
        if data <= self.data:
            if self.left:
                self.left.insert(data)
            else:
                self.left = TreeNode(data)
        else:
            if self.right:
                self.right.insert(data)
            else:
                self.right = TreeNode(data)


class BinarySearchTree:
    def __init__(self, tree_data):
        self._root = None
        for value in tree_data:
            self.insert(value)

    def insert(self, value):
        if self._root is None:
            self._root = TreeNode(value)
        else:
            self._root.insert(value)

    def data(self):
        return self._root

    def sorted_data(self):
        res = []

        def traverse(node):
            if node:
                traverse(node.left)
                res.append(node.data)
                traverse(node.right)

        traverse(self._root)
        return res